File: opportunity/manual_sources.py
"""Manual-source catalogue for BRIEF-FR-006 E23 (Track E, nodes E2+E3).

Every entry here corresponds to a `docs/SOURCE_REGISTRY.yaml` source whose dated recon
outcome is `manual_only` or `BLOCKED_POLICY`: automated reading is disallowed by robots,
terms, or credential-gating, or the source is a platform-application (tutoring) rather
than a postings feed. No adapter reads any of these. Work order C3 renders this
catalogue as the "Check manually" panel; this module only supplies the data and a
resolvable-deep-link test (`opportunity/test_manual_sources.py`).

Tutoring entries are `opportunity_type="platform_application"` and must never be
rendered as job postings -- they are registration/profile platforms, surfaced under
`track = tutoring` with a founder readiness checklist instead of posting fields.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from urllib.parse import quote

from opportunity.models import Track

# The founder's default search terms for prefilled deep links (data/ML/AI engineering).
FOUNDER_SEARCH_QUERY = "data engineer OR machine learning engineer OR AI engineer"


@dataclass(frozen=True, slots=True)
class ManualSource:
    source_id: str
    name: str
    track: Track
    opportunity_type: str  # "manual_only" | "platform_application"
    deep_link_template: str  # "{query}" is substituted with a URL-quoted search term, if present
    category: str  # "aggregator" | "regional" | "freelance" | "tutoring"
    policy_note: str = ""
    alert_route_available: bool = False
    alert_route_configured: bool = False
    readiness_checklist: tuple[str, ...] = field(default_factory=tuple)

    def deep_link(self, query: str = FOUNDER_SEARCH_QUERY) -> str:
        if "{query}" in self.deep_link_template:
            return self.deep_link_template.format(query=quote(query))
        return self.deep_link_template


_REDDIT_403_NOTE = (
    "Public JSON endpoint (https://www.reddit.com/r/<subreddit>.json) returned HTTP 403 "
    "on the first request (checked via r/forhire, 2026-09-03); AGENTS.md 403-stop rule "
    "observed, not retried, and generalized to the other six listed subreddits sharing "
    "the same host, endpoint pattern, and anti-bot policy rather than requesting each one "
    "individually against an already-blocking host."
)

TUTORING_READINESS_CHECKLIST: tuple[str, ...] = (
    "Founder profile photo and bio drafted",
    "Subject/skill areas selected",
    "Identity verification documents ready (platform-specific)",
    "Availability calendar defined",
    "Introductory/demo video recorded (where required)",
)

MANUAL_SOURCES: tuple[ManualSource, ...] = (
    # --- E2: aggregators and communities -----------------------------------------
    ManualSource(
        "hacker_news_who_is_hiring", "Hacker News \"Who is hiring?\"", Track.EMPLOYMENT,
        "manual_only", "https://hn.algolia.com/?q={query}&type=comment", "aggregator",
        "Registered here only as a fallback deep link; the primary route is the read-allowed "
        "adapter (see docs/SOURCE_REGISTRY.yaml: hacker_news_who_is_hiring, read=allowed).",
    ),
    ManualSource("reddit_forhire", "Reddit r/forhire", Track.EMPLOYMENT, "manual_only",
                 "https://old.reddit.com/r/forhire/search?q={query}&restrict_sr=on&sort=new", "aggregator", _REDDIT_403_NOTE),
    ManualSource("reddit_remotejobs", "Reddit r/remotejobs", Track.EMPLOYMENT, "manual_only",
                 "https://old.reddit.com/r/remotejobs/search?q={query}&restrict_sr=on&sort=new", "aggregator", _REDDIT_403_NOTE),
    ManualSource("reddit_machinelearningjobs", "Reddit r/MachineLearningJobs", Track.EMPLOYMENT, "manual_only",
                 "https://old.reddit.com/r/MachineLearningJobs/search?q={query}&restrict_sr=on&sort=new", "aggregator", _REDDIT_403_NOTE),
    ManualSource("reddit_datajobs", "Reddit r/datajobs", Track.EMPLOYMENT, "manual_only",
                 "https://old.reddit.com/r/datajobs/search?q={query}&restrict_sr=on&sort=new", "aggregator", _REDDIT_403_NOTE),
    ManualSource("reddit_hiring", "Reddit r/hiring", Track.EMPLOYMENT, "manual_only",
                 "https://old.reddit.com/r/hiring/search?q={query}&restrict_sr=on&sort=new", "aggregator", _REDDIT_403_NOTE),
    ManualSource("reddit_jobbit", "Reddit r/jobbit", Track.EMPLOYMENT, "manual_only",
                 "https://old.reddit.com/r/jobbit/search?q={query}&restrict_sr=on&sort=new", "aggregator", _REDDIT_403_NOTE),
    ManualSource("reddit_bigdatajobs", "Reddit r/bigdatajobs", Track.EMPLOYMENT, "manual_only",
                 "https://old.reddit.com/r/bigdatajobs/search?q={query}&restrict_sr=on&sort=new", "aggregator", _REDDIT_403_NOTE),
    ManualSource("ycombinator_work_at_a_startup", "Y Combinator Work at a Startup", Track.EMPLOYMENT, "manual_only",
                 "https://www.workatastartup.com/jobs?query={query}", "aggregator",
                 "robots.txt allows crawling (checked 2026-09-03) but no documented public JSON/RSS jobs API "
                 "was found within recon budget; full listings require an applicant account. Deep link only."),
    ManualSource("working_nomads", "Working Nomads", Track.EMPLOYMENT, "manual_only",
                 "https://www.workingnomads.com/jobs?search={query}", "aggregator",
                 "robots.txt allows crawling (checked 2026-09-03); guessed RSS/API paths "
                 "(/jobs.rss, /api/exposed_jobs/rss) returned HTTP 404. No documented public feed found. Deep link only."),
    ManualSource("remote_co", "Remote.co", Track.EMPLOYMENT, "manual_only",
                 "https://remote.co/remote-jobs/search/?search_keywords={query}", "aggregator",
                 "robots.txt fetch timed out on two attempts (2026-09-03), consistent with anti-bot mitigation "
                 "against a non-browser user agent; not retried a third time. Deep link only."),
    ManualSource("justremote", "JustRemote", Track.EMPLOYMENT, "manual_only",
                 "https://justremote.co/remote-jobs?search={query}", "aggregator",
                 "robots.txt allows crawling (checked 2026-09-03); no documented public JSON/RSS jobs API found. Deep link only."),
    ManualSource("wellfound", "Wellfound (AngelList Talent)", Track.EMPLOYMENT, "manual_only",
                 "https://wellfound.com/jobs?query={query}", "aggregator",
                 "Brief-designated alert-route source; job data API requires an authenticated account. "
                 "robots.txt checked 2026-09-03 (partial disallow on job/session paths).",
                 alert_route_available=True, alert_route_configured=False),
    ManualSource("arc_dev", "Arc.dev", Track.EMPLOYMENT, "manual_only",
                 "https://arc.dev/remote-jobs?search={query}", "aggregator",
                 "robots.txt allows crawling (checked 2026-09-03); no documented public jobs API found within recon budget. Deep link only."),
    ManualSource("ai_jobs_net", "ai-jobs.net", Track.EMPLOYMENT, "manual_only",
                 "https://ai-jobs.net/?s={query}", "aggregator",
                 "Recon inconclusive (checked 2026-09-03): robots.txt request returned an HTML SPA fallback "
                 "rather than a text robots file, and the candidate API path timed out. Not confirmed reachable "
                 "in this session; registered manual_only pending a future, successful recon pass."),
    ManualSource("otta", "Otta", Track.EMPLOYMENT, "manual_only",
                 "https://otta.com/jobs?q={query}", "aggregator",
                 "Brief-designated alert-route source; job data requires an authenticated account. "
                 "robots.txt checked 2026-09-03 (generic allow).",
                 alert_route_available=True, alert_route_configured=False),
    ManualSource("jobicy", "Jobicy", Track.EMPLOYMENT, "manual_only",
                 "https://jobicy.com/?s={query}", "aggregator",
                 "Already registered by prior recon: robots.txt returned HTTP 403 on all three attempts "
                 "(2026-09-02). Not re-probed per this order's facts. Deep link only."),

    # --- E2: regional and Arabic-language -----------------------------------------
    ManualSource("wuzzuf", "Wuzzuf", Track.EMPLOYMENT, "manual_only",
                 "https://wuzzuf.net/search/jobs/?q={query}", "regional",
                 "Alert-mailbox route already built (FR-004 inbox adapter); founder has not provided a mailbox "
                 "to configure. robots.txt uses Content-Signal directives, not classic Disallow (checked 2026-09-03).",
                 alert_route_available=True, alert_route_configured=False),
    ManualSource("bayt", "Bayt.com", Track.EMPLOYMENT, "manual_only",
                 "https://www.bayt.com/en/search/?q={query}", "regional",
                 "Alert-mailbox route already built; founder has not provided a mailbox to configure. "
                 "robots.txt explicitly disallows named bots (008, LinkedInBot, IndexBot) (checked 2026-09-03).",
                 alert_route_available=True, alert_route_configured=False),
    ManualSource("gulftalent", "GulfTalent", Track.EMPLOYMENT, "manual_only",
                 "https://www.gulftalent.com/jobs/search?keywords={query}", "regional",
                 "Alert-mailbox route already built; founder has not provided a mailbox to configure. "
                 "robots.txt explicitly disallows named bots (checked 2026-09-03).",
                 alert_route_available=True, alert_route_configured=False),
    ManualSource("naukrigulf", "Naukrigulf", Track.EMPLOYMENT, "manual_only",
                 "https://www.naukrigulf.com/jobs-in-uae?searchType=adv&keyword={query}", "regional",
                 "Alert-mailbox route already built; founder has not provided a mailbox to configure. "
                 "robots.txt fetch timed out on two attempts (2026-09-03); not retried a third time.",
                 alert_route_available=True, alert_route_configured=False),
    ManualSource("linkedin", "LinkedIn", Track.EMPLOYMENT, "manual_only",
                 "https://www.linkedin.com/jobs/search/?keywords={query}", "regional",
                 "robots.txt explicitly states automated access is strictly prohibited without express permission "
                 "(checked 2026-09-03). Alert-mailbox route available, unconfigured (no founder mailbox).",
                 alert_route_available=True, alert_route_configured=False),
    ManualSource("indeed", "Indeed", Track.EMPLOYMENT, "manual_only",
                 "https://www.indeed.com/jobs?q={query}", "regional",
                 "robots.txt permits generic SEO crawling but no public job-data API is documented; job alerts "
                 "require an account. Alert-mailbox route available, unconfigured (checked 2026-09-03).",
                 alert_route_available=True, alert_route_configured=False),

    # --- E3: freelance ---------------------------------------------------------
    ManualSource("mostaql", "Mostaql", Track.FREELANCE, "manual_only",
                 "https://mostaql.com/projects?filters%5Bkeyword%5D={query}", "freelance",
                 "robots.txt disallows /search* and /ajax/ but general browsing is not blocked (checked 2026-09-03); "
                 "no documented public JSON/RSS projects API found. Deep link only."),
    ManualSource("khamsat", "Khamsat", Track.FREELANCE, "manual_only",
                 "https://khamsat.com/community/services?filter%5Bkeyword%5D={query}", "freelance",
                 "robots.txt allows general browsing with several dynamic-query disallows (checked 2026-09-03); "
                 "no documented public API found. Deep link only."),
    ManualSource("contra", "Contra", Track.FREELANCE, "manual_only",
                 "https://contra.com/search/independents?q={query}", "freelance",
                 "robots.txt uses Content-Signal directives permitting search inclusion (ai-train=no, search=yes) "
                 "(checked 2026-09-03); no documented public gigs API found. Deep link only."),
    ManualSource("peopleperhour", "PeoplePerHour", Track.FREELANCE, "manual_only",
                 "https://www.peopleperhour.com/freelance-jobs?q={query}", "freelance",
                 "robots.txt allows general browsing (checked 2026-09-03); no documented public projects API found. Deep link only."),
    ManualSource("toptal", "Toptal", Track.FREELANCE, "manual_only",
                 "https://www.toptal.com/talent/apply", "freelance",
                 "Application-based per brief; Toptal has no browsable open project feed for unaccepted applicants. Deep link only."),
    ManualSource("upwork", "Upwork", Track.FREELANCE, "manual_only",
                 "https://www.upwork.com/nx/search/jobs/?q={query}", "freelance",
                 "robots.txt request itself returned HTTP 403 (checked 2026-09-03); host blocks non-browser access "
                 "outright, confirming the brief's note that the API is partner-only. Not retried; RSS search "
                 "feeds could not be safely reconned without a second request to an already-blocking host."),
    ManualSource("freelancer", "Freelancer.com", Track.FREELANCE, "manual_only",
                 "https://www.freelancer.com/jobs/?keyword={query}", "freelance",
                 "Stays credential-gated per brief instruction; existing registry entry (docs/SOURCE_REGISTRY.yaml) "
                 "left as-is. Deep link only."),

    # --- E3: tutoring (platform_application, never rendered as postings) -----------
    ManualSource("preply", "Preply", Track.TUTORING, "platform_application",
                 "https://preply.com/en/apply-to-preply-tutor", "tutoring",
                 "Platform application, not a postings feed (checked 2026-09-03).",
                 readiness_checklist=TUTORING_READINESS_CHECKLIST),
    ManualSource("superprof", "Superprof", Track.TUTORING, "platform_application",
                 "https://www.superprof.com/apply.html", "tutoring",
                 "Platform application, not a postings feed; robots.txt request itself returned HTTP 403 "
                 "(Cloudflare bot mitigation, checked 2026-09-03), confirming no automated route exists.",
                 readiness_checklist=TUTORING_READINESS_CHECKLIST),
    ManualSource("wyzant", "Wyzant", Track.TUTORING, "platform_application",
                 "https://www.wyzant.com/tutor/apply", "tutoring",
                 "Platform application, not a postings feed (checked 2026-09-03).",
                 readiness_checklist=TUTORING_READINESS_CHECKLIST),
    ManualSource("tutor_com", "Tutor.com", Track.TUTORING, "platform_application",
                 "https://www.tutor.com/apply", "tutoring",
                 "Platform application, not a postings feed (checked 2026-09-03).",
                 readiness_checklist=TUTORING_READINESS_CHECKLIST),
    ManualSource("chegg", "Chegg Tutors", Track.TUTORING, "platform_application",
                 "https://www.chegg.com/tutors/apply/", "tutoring",
                 "Platform application, not a postings feed (checked 2026-09-03).",
                 readiness_checklist=TUTORING_READINESS_CHECKLIST),
    ManualSource("cambly", "Cambly", Track.TUTORING, "platform_application",
                 "https://www.cambly.com/tutor", "tutoring",
                 "Platform application, not a postings feed (checked 2026-09-03).",
                 readiness_checklist=TUTORING_READINESS_CHECKLIST),
)


def by_category(category: str) -> tuple[ManualSource, ...]:
    return tuple(item for item in MANUAL_SOURCES if item.category == category)


def tutoring_platform_cards() -> tuple[ManualSource, ...]:
    return tuple(item for item in MANUAL_SOURCES if item.track == Track.TUTORING and item.opportunity_type == "platform_application")

File: web/lib/data/manual-sources.ts
/**
 * E23 UI half — the "Check manually" panel's data.
 *
 * BRIEF-FR-006 C5 UPDATE: `GET /api/manual-sources` now exists
 * (`api/routes_api.py::manual_sources_route`) and serves
 * `opportunity/manual_sources.py::MANUAL_SOURCES` directly, read-only, with
 * zero outbound network I/O (`api/test_api.py::ManualSourcesRouteTest`
 * asserts that). This file's static transcription is corrected against the
 * live module as of this order (the stale `hacker_news_who_is_hiring`
 * entry — removed from the Python catalogue by a prior council review
 * because it is a fully automated, read-allowed adapter, not a manual
 * fallback — is deleted here too), but it is still a duplicate, not the
 * live route: `components/feed/manual-sources-panel.tsx` (the only
 * consumer) is outside this order's allowed-files list and still imports
 * `MANUAL_SOURCES` as a synchronous array, so switching this file to an
 * async `fetch("/api/manual-sources")` would require also converting that
 * component to load state asynchronously — out of scope here. A future
 * order should do both together and delete this file's static array.
 *
 * Deliberately excludes every `opportunity_type: "platform_application"`
 * (tutoring) entry: the Python module's own docstring says those "must
 * never be rendered as job postings" and belong to a founder-readiness
 * checklist feature this order does not build.
 */

export type ManualSourceCategory = "aggregator" | "regional" | "freelance"

export interface ManualSource {
  sourceId: string
  name: string
  track: "employment" | "freelance"
  category: ManualSourceCategory
  deepLink: string
  policyNote: string
}

const FOUNDER_SEARCH_QUERY =
  "data engineer OR machine learning engineer OR AI engineer"

function deepLink(template: string, query: string = FOUNDER_SEARCH_QUERY): string {
  return template.includes("{query}")
    ? template.replace("{query}", encodeURIComponent(query))
    : template
}

const REDDIT_403_NOTE =
  "Public JSON endpoint returned HTTP 403 on the first request (checked via r/forhire, 2026-09-03); AGENTS.md 403-stop rule observed, not retried."

export const MANUAL_SOURCES: ManualSource[] = [
  // --- aggregators and communities ---
  {
    sourceId: "reddit_forhire",
    name: "Reddit r/forhire",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink(
      "https://old.reddit.com/r/forhire/search?q={query}&restrict_sr=on&sort=new"
    ),
    policyNote: REDDIT_403_NOTE,
  },
  {
    sourceId: "reddit_remotejobs",
    name: "Reddit r/remotejobs",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink(
      "https://old.reddit.com/r/remotejobs/search?q={query}&restrict_sr=on&sort=new"
    ),
    policyNote: REDDIT_403_NOTE,
  },
  {
    sourceId: "reddit_machinelearningjobs",
    name: "Reddit r/MachineLearningJobs",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink(
      "https://old.reddit.com/r/MachineLearningJobs/search?q={query}&restrict_sr=on&sort=new"
    ),
    policyNote: REDDIT_403_NOTE,
  },
  {
    sourceId: "reddit_datajobs",
    name: "Reddit r/datajobs",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink(
      "https://old.reddit.com/r/datajobs/search?q={query}&restrict_sr=on&sort=new"
    ),
    policyNote: REDDIT_403_NOTE,
  },
  {
    sourceId: "reddit_hiring",
    name: "Reddit r/hiring",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink(
      "https://old.reddit.com/r/hiring/search?q={query}&restrict_sr=on&sort=new"
    ),
    policyNote: REDDIT_403_NOTE,
  },
  {
    sourceId: "reddit_jobbit",
    name: "Reddit r/jobbit",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink(
      "https://old.reddit.com/r/jobbit/search?q={query}&restrict_sr=on&sort=new"
    ),
    policyNote: REDDIT_403_NOTE,
  },
  {
    sourceId: "reddit_bigdatajobs",
    name: "Reddit r/bigdatajobs",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink(
      "https://old.reddit.com/r/bigdatajobs/search?q={query}&restrict_sr=on&sort=new"
    ),
    policyNote: REDDIT_403_NOTE,
  },
  {
    sourceId: "ycombinator_work_at_a_startup",
    name: "Y Combinator Work at a Startup",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink("https://www.workatastartup.com/jobs?query={query}"),
    policyNote:
      "robots.txt allows crawling but no documented public JSON/RSS jobs API was found; full listings require an applicant account.",
  },
  {
    sourceId: "working_nomads",
    name: "Working Nomads",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink("https://www.workingnomads.com/jobs?search={query}"),
    policyNote:
      "robots.txt allows crawling; guessed RSS/API paths returned HTTP 404. No documented public feed found.",
  },
  {
    sourceId: "remote_co",
    name: "Remote.co",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink(
      "https://remote.co/remote-jobs/search/?search_keywords={query}"
    ),
    policyNote:
      "robots.txt fetch timed out on two attempts, consistent with anti-bot mitigation; not retried a third time.",
  },
  {
    sourceId: "justremote",
    name: "JustRemote",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink("https://justremote.co/remote-jobs?search={query}"),
    policyNote:
      "robots.txt allows crawling; no documented public JSON/RSS jobs API found.",
  },
  {
    sourceId: "wellfound",
    name: "Wellfound (AngelList Talent)",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink("https://wellfound.com/jobs?query={query}"),
    policyNote:
      "Brief-designated alert-route source; job data API requires an authenticated account.",
  },
  {
    sourceId: "arc_dev",
    name: "Arc.dev",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink("https://arc.dev/remote-jobs?search={query}"),
    policyNote:
      "robots.txt allows crawling; no documented public jobs API found within recon budget.",
  },
  {
    sourceId: "ai_jobs_net",
    name: "ai-jobs.net",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink("https://ai-jobs.net/?s={query}"),
    policyNote:
      "Recon inconclusive: robots.txt request returned an HTML SPA fallback; not confirmed reachable this session.",
  },
  {
    sourceId: "otta",
    name: "Otta",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink("https://otta.com/jobs?q={query}"),
    policyNote:
      "Brief-designated alert-route source; job data requires an authenticated account.",
  },
  {
    sourceId: "jobicy",
    name: "Jobicy",
    track: "employment",
    category: "aggregator",
    deepLink: deepLink("https://jobicy.com/?s={query}"),
    policyNote:
      "robots.txt returned HTTP 403 on all three attempts. Not re-probed.",
  },

  // --- regional and Arabic-language ---
  {
    sourceId: "wuzzuf",
    name: "Wuzzuf",
    track: "employment",
    category: "regional",
    deepLink: deepLink("https://wuzzuf.net/search/jobs/?q={query}"),
    policyNote:
      "Alert-mailbox route already built (FR-004 inbox adapter); founder has not provided a mailbox to configure.",
  },
  {
    sourceId: "bayt",
    name: "Bayt.com",
    track: "employment",
    category: "regional",
    deepLink: deepLink("https://www.bayt.com/en/search/?q={query}"),
    policyNote:
      "Alert-mailbox route already built; founder has not provided a mailbox to configure.",
  },
  {
    sourceId: "gulftalent",
    name: "GulfTalent",
    track: "employment",
    category: "regional",
    deepLink: deepLink("https://www.gulftalent.com/jobs/search?keywords={query}"),
    policyNote:
      "Alert-mailbox route already built; founder has not provided a mailbox to configure.",
  },
  {
    sourceId: "naukrigulf",
    name: "Naukrigulf",
    track: "employment",
    category: "regional",
    deepLink: deepLink(
      "https://www.naukrigulf.com/jobs-in-uae?searchType=adv&keyword={query}"
    ),
    policyNote:
      "Alert-mailbox route already built; founder has not provided a mailbox to configure.",
  },
  {
    sourceId: "linkedin",
    name: "LinkedIn",
    track: "employment",
    category: "regional",
    deepLink: deepLink("https://www.linkedin.com/jobs/search/?keywords={query}"),
    policyNote:
      "robots.txt explicitly states automated access is strictly prohibited without express permission.",
  },
  {
    sourceId: "indeed",
    name: "Indeed",
    track: "employment",
    category: "regional",
    deepLink: deepLink("https://www.indeed.com/jobs?q={query}"),
    policyNote:
      "robots.txt permits generic SEO crawling but no public job-data API is documented; job alerts require an account.",
  },

  // --- freelance ---
  {
    sourceId: "mostaql",
    name: "Mostaql",
    track: "freelance",
    category: "freelance",
    deepLink: deepLink(
      "https://mostaql.com/projects?filters%5Bkeyword%5D={query}"
    ),
    policyNote:
      "robots.txt disallows /search* and /ajax/ but general browsing is not blocked; no documented public JSON/RSS projects API found.",
  },
  {
    sourceId: "khamsat",
    name: "Khamsat",
    track: "freelance",
    category: "freelance",
    deepLink: deepLink(
      "https://khamsat.com/community/services?filter%5Bkeyword%5D={query}"
    ),
    policyNote:
      "robots.txt allows general browsing with several dynamic-query disallows; no documented public API found.",
  },
  {
    sourceId: "contra",
    name: "Contra",
    track: "freelance",
    category: "freelance",
    deepLink: deepLink("https://contra.com/search/independents?q={query}"),
    policyNote:
      "robots.txt uses Content-Signal directives permitting search inclusion; no documented public gigs API found.",
  },
  {
    sourceId: "peopleperhour",
    name: "PeoplePerHour",
    track: "freelance",
    category: "freelance",
    deepLink: deepLink("https://www.peopleperhour.com/freelance-jobs?q={query}"),
    policyNote:
      "robots.txt allows general browsing; no documented public projects API found.",
  },
  {
    sourceId: "toptal",
    name: "Toptal",
    track: "freelance",
    category: "freelance",
    deepLink: "https://www.toptal.com/talent/apply",
    policyNote:
      "Application-based per brief; Toptal has no browsable open project feed for unaccepted applicants.",
  },
  {
    sourceId: "upwork",
    name: "Upwork",
    track: "freelance",
    category: "freelance",
    deepLink: deepLink("https://www.upwork.com/nx/search/jobs/?q={query}"),
    policyNote:
      "robots.txt request itself returned HTTP 403; host blocks non-browser access outright.",
  },
  {
    sourceId: "freelancer",
    name: "Freelancer.com",
    track: "freelance",
    category: "freelance",
    deepLink: deepLink("https://www.freelancer.com/jobs/?keyword={query}"),
    policyNote: "Stays credential-gated per brief instruction.",
  },
]
